README.md

```diff
@@ -1 +1,544 @@
+
+The `HTTPHeaders` type represents HTTP headers.
+
+```swift
+public typealias HTTPHeaders = [HTTPHeader]
+```
+
+### Motivation
+
+Unfortunately the HTTP protocol allows duplicate headers (see `Set-Cookie`). So we have to make it an array instead of a dictionary.
+
+## HTTPBody
+
+The `HTTPBody` type represents an HTTP body in buffer or stream form.
+
+```swift
+public enum HTTPBody {
+    case BufferBody(Data)
+    case StreamBody(Stream)
+}
+```
+
+### Motivation
+
+Buffer bodies are best suited for HTTP messages with `Content-Length` headers. Stream bodies are best suited for HTTP messages with `Transfer-Encoding` set to `chunked`.
+
+## Storage
+
+The `Storage` type represents arbitrary data that can be passed between middleware and a responder in a chain.
+
+```swift
+public typealias Storage = [String: Any]
+```
+
+### Motivation
+
+Having a storage in the HTTP messages makes middleware much more useful since they can pass forward parsed/serialized content, session information or just about any custom data.
+
+## HTTPMessage
+
+The `HTTPMessage` protocol represents properties common to HTTP messages (request/response).
+
+```swift
+public protocol HTTPMessage {
+    var version: HTTPVersion { get set }
+    var headers: HTTPHeaders { get set }
+    var body: HTTPBody { get set }
+    var storage: Storage { get set }
+}
+```
+
+### Motivation
+
+Having the `HTTPMessage` protocol makes adding computed properties common to requests and responses more convenient. It removes code duplication.
+
+## HTTPMethod
+
+The `HTTPMethod` type represents an HTTP method.
+
+```swift
+public enum HTTPMethod {
+    case Delete
+    case Get
+    case Head
+    case Post
+    case Put
+    case Connect
+    case Options
+    case Trace
+    case Copy
+    case Lock
+    case MkCol
+    case Move
+    case PropFind
+    case PropPatch
+    case Search
+    case Unlock
+    case Bind
+    case Rebind
+    case Unbind
+    case ACL
+    case Report
+    case MkActicity
+    case Checkout
+    case Merge
+    case MSearch
+    case Notify
+    case Subscribe
+    case Unsubscribe
+    case Patch
+    case Purge
+    case MkCalendar
+    case Link
+    case Unlink
+    case Raw(method: String)
+}
+```
+
+### Motivation
+
+Having `HTTPMethod` as an enum with a `Raw` case makes it convenient and flexible at the same time.
+
+## URI
+
+The `URI` type represents an URI.
+
+```swift
+public struct URI {
+    public struct UserInfo {
+        public var username: String
+        public var password: String
+    }
+
+    public var scheme: String?
+    public var userInfo: UserInfo?
+    public var host: String?
+    public var port: Int?
+    public var path: String?
+    public var query: [String: String]
+    public var fragment: String?
+}
+```
+
+### Motivation
+
+Having `URI` as a `struct` allows it to conform to protocols in extensions.
+
+## HTTPRequest
+
+The `HTTPRequest` type represents an HTTP request.
+
+```swift
+public struct HTTPRequest: HTTPMessage {
```

Comment:
> Why couldn't you just have protocol extensions that conform to CustomStringConvertible, etc. Then the frameworks wouldn't have to implement that?

Yeah we could have that. And frameworks could extend them if they want.

> Also what about Kitura would "simply not work"?

This is just conjecture, but I think It wouldn't work because the request depends on:

`BlueSocketWriter`
`CurlInvokerDelegate`

If request is a protocol, kitura's request wouldn't be able to conform to that protocol. And it seems that these protocols are deep rooted in kitura's architecture. But maybe it could be possible to get rid of them or wrap a request that conforms to `S4.HTTPRequest` inside kitura's request. It surely wouldn't be as easy as in the other frameworks.